Index: network-manager-0.9.10.0/src/settings/plugins/ofono/parser.c
===================================================================
--- network-manager-0.9.10.0.orig/src/settings/plugins/ofono/parser.c
+++ network-manager-0.9.10.0/src/settings/plugins/ofono/parser.c
@@ -53,8 +53,6 @@ ofono_update_connection_from_context (NM
 	char *idstr = NULL;
 	char *uuid_base = NULL;
 	char *uuid = NULL;
-	const char *username = NULL;
-	const char *password = NULL;
 
 	s_con = nm_connection_get_setting_connection (connection);
 	if(!s_con) {
@@ -91,25 +89,6 @@ ofono_update_connection_from_context (NM
 	 */
 	g_object_set (s_gsm, NM_SETTING_GSM_NUMBER, "*99#", NULL);
 
-	/* Set the APN we got from oFono config. */
-	g_object_set (s_gsm,
-	              NM_SETTING_GSM_APN, g_hash_table_lookup (context, "Apn"),
-	              NULL);
-
-	username = g_hash_table_lookup (context, "Username");
-	if (username && g_strcmp0 (username, "") != 0) {
-        	g_object_set (s_gsm,
-		              NM_SETTING_GSM_USERNAME, username,
-	                      NULL);
-	}
-
-	password = g_hash_table_lookup (context, "Password");
-	if (password && g_strcmp0 (password, "") != 0) {
-        	g_object_set (s_gsm,
-		              NM_SETTING_GSM_PASSWORD, password,
-	                      NULL);
-	}
-
 	nm_log_info (LOGD_SETTINGS, "SCPlugin-Ofono: "
 	             "update_connection_setting_from_context: name:%s, path:%s, type:%s, id:%s, uuid: %s",
 	             (char *) g_hash_table_lookup (context, "Name"),
